### PR TITLE
[CSM-549] Proof of concept of serving a V1 endpoint with V0 logic

### PR DESCRIPTION
**Pull request by adinapoli-iohk**
This not-at-all-polished PR ships a proof-of-concept showing how we can effectively serve a V1 endpoint with V0 logic. It boils down to essentially re-use the V0 natural transformation and adding a `Migration` module which is just a eventually-to-be-removed shim which will help us converting back and forth from the various datatypes.

Most of the datatypes are stubbed, and the `MonadV1` is essentially just a very constrained type alias for `WalletWebMode`, but hopefully in the future we can have our own monadic stack which can in principle diverge from V0 (if needed).

*Most importantly*, this PR also moves the `BackupPhrase` module from `cardano-sl` to `cardano-sl-wallet`, where it should belong (cc @gromakovsky , this will hopefully land first into `alfredo/wallet-new` and later into `master`). You will notice how I had to also move around a bunch of test modules to also move the `BackupPhrase` tests run in `cardano-sl-wallet`. If you have ideas how to make this cleaner, let me know.

@flyingleafe Do you think you can give this a quick review and tell me what you think about the general approach? I'm also very happy if this doesn't make any sense to you and we can surely discuss improvements, but I would capture that in a separate ticket so that we don't block merging this PR which still ships valuable refactoring I would like to see landing on `master` by the end of the week, so that we don't duplicate work with the core team 😉 

Thank you guys!

### CODE REVIEW

File: lib/src/Pos/Client/CLI/NodeOptions.hs [hunk 1]
```diff
@@ -13,33 +13,31 @@ module Pos.Client.CLI.NodeOptions
...
-import           Pos.Util.BackupPhrase (BackupPhrase, backupPhraseWordsNum)
-import           Pos.Util.CompileInfo (CompileTimeInfo (..), HasCompileInfo, compileInfo)
-import           Pos.Util.TimeWarp (NetworkAddress)
+import           Pos.Client.CLI.Options       (CommonArgs (..), commonArgsParser, optionalJSONPath)
+import           Pos.HealthCheck.Route53      (route53HealthCheckOption)
+import           Pos.Network.CLI              (NetworkConfigOpts, networkConfigOption)
+import           Pos.Statistics               (EkgParams, StatsdParams, ekgParamsOption,
+                                               statsdParamsOption)
+import           Pos.Util.CompileInfo         (CompileTimeInfo (..), HasCompileInfo, compileInfo)
+import           Pos.Util.TimeWarp            (NetworkAddress)
```
Comment (gromakovsky) [inline lib/src/Pos/Client/CLI/NodeOptions.hs]: It looks like `stylish-haskell` picked wrong configuration (the old one).

File: wallet-new/src/Cardano/Wallet/API/V1/Migration.hs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+walletAssurance :: Iso' V0.CWalletAssurance V1.AssuranceLevel
+walletAssurance = iso migrate migrate
+
+--
+instance Migrate V0.CPassPhrase V1.SpendingPassword where
+  migrate (V0.CPassPhrase pp) = pp
+
+instance Migrate V1.SpendingPassword V0.CPassPhrase where
+  migrate pp = V0.CPassPhrase pp
```
Comment (gromakovsky) [inline wallet-new/src/Cardano/Wallet/API/V1/Migration.hs]: In some places indentation is 4 spaces, in some places it's 2 spaces. Can you stick to 4 spaces (as per style-guide) please?

File: lib/src/Pos/Client/CLI/NodeOptions.hs (hunk 1 above)
Comment (adinapoli-iohk) [reply lib/src/Pos/Client/CLI/NodeOptions.hs]: Yes, I have been wondering about this for a while, and I have still to understand why, considering the correct stylish-haskell is now committed to the main repo. I will have to double-check.

File: wallet-new/src/Cardano/Wallet/API/V1/Migration.hs (hunk 2 above)
Comment (adinapoli-iohk) [reply wallet-new/src/Cardano/Wallet/API/V1/Migration.hs]: Is there a way to enforce this with stylish-haskell as well? I come from projects where we were using 2 spaces (which, btw, allow more test to be on one line, so that's surprising why you guys are using 4 spaces), so I will keep forgetting as muscle memory will kick in. 😢

**Comment by adinapoli-iohk** [comment]
> Maybe it makes sense to move all tests into single package which will depend on all other packages we want to test.

Yea, that was something I thought about, but looks like something too ambitious to do here as part of this PR, but I'm happy to discuss with the team plans on this front for the near future.

File: wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs [hunk 3]
```diff
@@ -2,30 +2,67 @@ module Cardano.Wallet.API.V1.Handlers.Wallets where
...
+mkPassPhrase text = do
+  bs <- Base16.decode text
+  let bl = BS.length bs
+  -- Currently passphrase may be either 32-byte long or empty (for
+  -- unencrypted keys).
+  if bl == 0 || bl == passphraseLength
+      then pure $ ByteArray.convert bs
+      else fail . toString $ sformat
+           ("Expected password length 0 or "%int%", not "%int)
+           passphraseLength bl
```
Comment (flyingleafe) [inline wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs]: I'm strongly for getting rid of proxy types like `SpendingPassword` completely and instead providing proper JSON instances for target datatypes like `PassPhrase`, and including validation logic in these JSON instances. Proxy types (like `CWallet`, etc. in old wallet) are pain and they were initially introduced to make autogenerated code for purescript-bridge possible, but we're not going to use purescript-bridge with API V1 anymore.

File: wallet-new/server/Cardano/Wallet/Server.hs [hunk 4]
```diff
@@ -19,12 +21,15 @@ import           Servant
...
-    :: forall ctx m.
-       ( MonadFullWalletWebMode ctx m )
-    => (forall a. m a -> Handler a)
-    -> Server WalletAPI
+walletServer :: ( Migration.HasConfiguration
+                , Migration.HasCompileInfo
+                , Migration.HasUpdateConfiguration
+                , Migration.HasSscConfiguration
+                , Migration.HasNodeConfiguration
+                , Migration.HasInfraConfiguration
```
Comment (flyingleafe) [inline wallet-new/server/Cardano/Wallet/Server.hs]: We combined type constraint `HasConfigurations` for this. It's in `cardano-sl`, module `Pos.Launcher.Configuration`.

File: wallet-new/src/Cardano/Wallet/API/V1/Migration.hs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+--
+-- Instances
+--
+
+instance Migrate V0.CWallet V1.Wallet where
+    migrate V0.CWallet{..} =
+        V1.Wallet { V1.walId   = migrate @(V0.CId V0.Wal) cwId
+                  , V1.walName = V0.cwName cwMeta
+                  , V1.walBalance = migrate @V0.CCoin @V1.RenderedBalance cwAmount
```
Comment (flyingleafe) [inline wallet-new/src/Cardano/Wallet/API/V1/Migration.hs]: Just wondering: doesn't GHC  deduce these types automatically given type of `cwAmount` and `walBalance`? I believe it would be less verbose to not write these type applications, if it does.

You will probably say that explicit type applications help to understand quicker which exact `Migrate` instances are used here, but I don't think that this effect is significant (though it's arguable). It's minor anyway.

File: wallet-new/src/Cardano/Wallet/API/V1/Migration.hs (hunk 2 above)
Comment (flyingleafe) [reply wallet-new/src/Cardano/Wallet/API/V1/Migration.hs]: @adinapoli-iohk AFAIK, unfortunately, no, there's no way to control indentation with stylish-haskell. If you're using Emacs though, I can share my `.dir-locals.el` for this project, which sets up indentation accordingly to style guide.

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs [hunk 6]
```diff
@@ -245,46 +262,57 @@ type Coins = Int
...
-  arbitrary = WalletUpdate <$> fmap getPositive arbitrary
-                           <*> arbitrary
-                           <*> fmap fromString arbitrary
+  arbitrary = WalletUpdate <$> arbitrary
+                           <*> pure "My Wallet"
+
+-- A 'RenderedBalance' represent the @rendered@ UTXO for
+-- this 'Wallet'.
+newtype RenderedBalance = RenderedBalance { renderBalance :: Text }
+                        deriving (Show, Eq, Generic)
```
Comment (flyingleafe) [inline wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: Comment is incorrect, I believe you wanted to say not "UTXO", but "balance".
Also, as I already said, I'm really against using such proxy types in new wallet. I'm really for simply using `Coin` here (and making a JSON instance which parses `Coin` from text). I don't see any particular reason we should store and manipulate balance as simple `Text` on backend side -- if you do see, please enlighten me.

File: wallet-new/src/Cardano/Wallet/Orphans.hs [hunk 7]
```diff
@@ -0,0 +1,19 @@
+{- |
+   An orphanage for all the non-interesting orphan instances.
+-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module Cardano.Wallet.Orphans where
```
Comment (flyingleafe) [inline wallet-new/src/Cardano/Wallet/Orphans.hs]: I believe that you can actually name this `Cardano.Wallet.Aeson` and put here all Aeson instances which wallet-new uses. We do so with `Bi` instances in Core, we do so with JSON instances in old wallet and I still think that this is a good approach.

File: wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs (hunk 3 above)
Comment (adinapoli-iohk) [reply wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs]: > providing proper JSON instances for target datatypes like PassPhrase, and including validation logic in these JSON instances

I see your point but I tend to disagree just because I think it's not their place. JSON instances should do just one thing well, which is making sure the input data is valid JSON, nothing else. Overloading them with too many responsibility feels like the wrong design choice. What do you think? 😉 

I do see why you find this appealing, which is because this way your handlers look nice and clean, but I would much rather create some extra layer in our handlers to perform the data validation than overloading very specialised instances with too many responsibilities.

File: wallet-new/server/Cardano/Wallet/Server.hs (hunk 4 above)
Comment (adinapoli-iohk) [reply wallet-new/server/Cardano/Wallet/Server.hs]: Oh, neat, I had no clue we had such a thing!

File: wallet-new/src/Cardano/Wallet/API/V1/Migration.hs (hunk 5 above)
Comment (adinapoli-iohk) [reply wallet-new/src/Cardano/Wallet/API/V1/Migration.hs]: You are probably right, GHC should be smart enough and as far as we are concerned this module will be removed eventually, so I'm happy either way. I will try to remove the type annotation, and we will re-assess later on if that compromised readability too much.

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs (hunk 6 above)
Comment (adinapoli-iohk) [reply wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: > Comment is incorrect, I believe you wanted to say not "UTXO", but "balance".

I will bite the bullet and look like a noob (which I am 😂 ): isn't - practically speaking - the UTXO for a wallet also equivalent to its balance?  

> Also, as I already said, I'm really against using such proxy types in new wallet. 

I hear you, but I think at the same time there are situation when we cannot avoid them, because we have an _impedance mismatch_ between the client types and the backend types, which is what you guys learnt the hard way with the old wallet (cfr. the proliferation of C-prefixed types, like `CCoin` & Co.

I feel there is a tension in the design space we will need to investigate; from one side we don't want all this extra types, but on the other we cannot overload our `Aeson` instances with too much logic (but maybe I'm wrong and you can convince me otherwise 😀 ). But for sure there are situation when we cannot avoid these client types, for example when creating or updating a wallet (for example). Let's discuss!

File: wallet-new/src/Cardano/Wallet/Orphans.hs (hunk 7 above)
Comment (adinapoli-iohk) [reply wallet-new/src/Cardano/Wallet/Orphans.hs]: I think that's a good suggestion. As long as we have a uniform scheme for dealing with that, I'm fine sticking to well-throttled conventions.

File: wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs (hunk 3 above)
Comment (dcoutts) [reply wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs]: I just had a quick chat with @adinapoli-iohk . I explained that my view is that one should probably decide this on a case by case basis. If the external syntax & validation is fixed and the validation is cheap then having just the one type is fine. If the validation is expensive and one needs to be able to read without validating, or if there are multiple different external syntaxes, or the validation is complex or needs additional input, then a two-step approach is better. For example with decoding blocks from the network we want a two-step process, to avoid the network layer having to do the expensive validation, and have that be done only by the logic layer.

Comment (flyingleafe) [reply wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs]: @dcoutts I completely agree, and I think that such types as `Coin`, `PassPhrase` and `Address` go to the first category, Reasons:
- Validation for them is simple and cheap (for `Address` there's actually no validation in JSON instance directly, because JSON representation of address is its binary representation encoded in base58, so we simply reuse validation logic from `Bi` instance)
- We never need to access textual representation of these types directly in code (only to log them, but we have `Buildable` instances for that)

I cannot actually come up right away with any type which is transferred via HTTP API and which goes into second category. 

Regarding your words, @adinapoli-iohk:
> JSON instances should do just one thing well, which is making sure the input data is valid JSON, nothing else.

No, JSON instances should make sure that input data _is a valid JSON representation of value of given type_. Consider the following example: say you have this datatype:
```
data Foo = { bar :: Int, baz :: String }
```

Now your endpoint accepts `Foo` as a request body. Then you have a query with a following request body:
```
{}
```
Or this one:
```
{"patak": {"bardaq": "skovoroda"}}
```
Are these inputs valid JSON? Definitely. Does it mean that JSON instance for `Foo` which tries to construct `Foo` object from these JSON inputs shouldn't throw an error? No, because a valid `Foo` object cannot be meaningfully constructed given either of these JSON inputs.

It's the same with such types as `PassPhrase` or `Address`: JSON instance need to check whether or not given JSON represents valid value of these datatypes -- nothing else.

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs (hunk 6 above)
Comment (flyingleafe) [reply wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: Well, UTXO is an abbreviation for "Unspent Transaction Outputs", which means that UTXO denotes a set of unspent transaction outputs (surprisingly enough). Unspent transaction output is a pair `(TxIn, TxOut)`, where `TxIn` is basically a pair of transaction hash and index of given output in this transaction, and `TxOut` is a pair of output address and amount of money associated with this output. Balance is simply an unspent amount of money on address, so it's a single `Coin` value. They are not equivalent -- you count balance if you know Utxo, but not the other way round.

Continuing this argument about proxy types:
- Please read my comment after Duncan's comment above
- I'm not sure what do you mean by _impedance mismatch_ here.
- C-prefixed types was introduced in order to generate Purescript types with similar JSON instances automatically via `purescript-bridge`. We couldn't do this for core types, e.g. `Address`, because it made no sense to introduce all the complicated internal structure of actual `Address` datatype to the client. And I think that this was a _bad_ design decision. Frontend guys were pissed off by purescript bridge, and we were suffering because of all this damned C-types. As soon as we're getting rid of purescript bridge, there's no more reason to bear with suffering from proxy types anymore.

File: wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs (hunk 3 above)
Comment (adinapoli-iohk) [reply wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs]: @flyingleafe First of all, thanks for the nice reply. I think a healthy debate is always the best way to reach the best design possible. Let me start from the bottom! 

> JSON instances should make sure that input data is a valid JSON representation of value of given type.

Yes, and I'm afraid I have spectacularly failed in my attempt to explain myself. What you say is correct, but what I meant was that one may argue (in some cases) that your input type should respect the "type footprint" without doing any validation inside. Let's take the example you posted: we all agree that those two raw JSONs shouldn't be reified into `Foo`, but if you give me `{"bar": 42, "baz": "foo" }` now it's a matter of whether or not you want to embed special validation logic to reject this input because, say, `Foo` should have `bar` always less than 50. So I hope now is a bit clearer why a was a bit on the fence in allowing validation in our Aeson instances, and that's because there is always the risk of restricting the domain of our types too much.

After speaking with @dcoutts though, I also see the other side of the coin, which is that these Aeson instances are really parsers, and parsers usually come with semantic attached to them (for example, if I was rolling the `FromJSON PassPhrase` parser by hand, I would have surely had some parsing rule to check if the passphrase length was 32 bits long, so that's essentially the same thing that you are proposing here).

> I cannot actually come up right away with any type which is transferred via HTTP API and which goes into second category.

I can think of one! 😉  That would be a `BackupPhrase`. Whilst we might want to embed validation logic into our `FromJSON` instance to check that the words are BIP-39 compliant, surely we don't want to embed validation logic on its _length_, because that might change in the future due to requirements; today is 12 words long, for an external wallet can be 24, for other cases can be 15. By embedding _too much_ validation logic we are constraining the subset of valid types too much, imho. Nothing which cannot be overcome by newtypes, mind you, but I still think this is an example of where we might want to check some invariant but not necessarily all of them.

The TL;DR is that I'm on board with the validation logic into `Aeson` instances, provided we do this with caution and we assess how much to validate on a case-by-case  basis. Do you think you can leave with that? 😉

Comment (flyingleafe) [reply wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs]: @adinapoli-iohk I agree with the example of `BackupPhrase`, and I'm glad that we've seemingly reached the consensus here :)

Comment (adinapoli-iohk) [reply wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs]: Of course we are! :) Ok, leave with me, I will amen this PR accordingly with all we have said here so far.

**Comment by adinapoli-iohk** [comment]
@flyingleafe I think this one now ticks all your quality boxes. Something I would like us to investigate is how we can generate meaningful examples for the swagger specs for secret stuff like a `PassPhrase`. Due to the fact the `Show` and `Buildable` instances for a `PassPhrase` does not show the secret (quite rightly), we cannot use the arbitrarily-generated value as-it-is, because it will be shown this way in Swagger:

screen shot 2017-11-16 at 15 53 36

Which means a user cannot really test the API in Swagger because the provided example won't work. Probably the solution would be to tweak somehow the `ToDocs` instance for the type.
